File: examples/Python/rtpapa.py
# encoding: utf-8
#
#   Custom routing Router to Papa (ROUTER to REP)
#
#   Author: Jeremy Avnet (brainsik) <spork(dash)zmq(at)theory(dot)org>
#

import time

import zmq

import zhelpers


context = zmq.Context()
client = context.socket(zmq.ROUTER)
client.bind("ipc://routing.ipc")

worker = context.socket(zmq.REP)
worker.setsockopt(zmq.IDENTITY, "A")
worker.connect("ipc://routing.ipc")

# Wait for sockets to stabilize
time.sleep(1)

client.send("A", zmq.SNDMORE)
client.send("address 3", zmq.SNDMORE)
client.send("address 2", zmq.SNDMORE)
client.send("address 1", zmq.SNDMORE)
client.send("", zmq.SNDMORE)
client.send("This is the workload")

# Worker should get just the workload
zhelpers.dump(worker)

# We don't play with envelopes in the worker
worker.send("This is the reply")

# Now dump what we got off the ROUTER socket...
zhelpers.dump(client)

File: examples/Haskell/rtpapa.hs
module Main where

import System.ZMQ
import ZHelpers (dumpSock)

import Control.Concurrent (threadDelay)
import Data.ByteString.Char8 (pack)

main :: IO ()
main = withContext 1 $ \context -> do
    withSocket context XRep $ \client -> do
        bind client "ipc://routing.ipc"
        withSocket context Rep $ \worker -> do
            setOption worker (Identity "A")
            connect worker "ipc://routing.ipc"
            
            threadDelay $ 2 * 1000 * 1000
            send client (pack "A") [SndMore]
            send client (pack "address 3") [SndMore]
            send client (pack "address 2") [SndMore]
            send client (pack "address 1") [SndMore]
            send client (pack "") [SndMore]
            send client (pack "This is the workload") []

            dumpSock worker
            send worker (pack "This is the reply") []
            
            dumpSock client